addon/file.ts

```diff
@@ -0,0 +1,323 @@
+import { tracked } from '@glimmer/tracking';
+
+import FileReader from './system/file-reader';
+import { upload } from './system/upload';
+import uuid from './system/uuid';
+
+import Queue from './queue';
+
+/**
+ * Possible file states.
+ *
+ * @remarks
+ *
+ * Here is the statechart describing the flow of state:
+ *
+ * ```
+ *       .------.     .---------.     .--------.
+ *   o--| queued |-->| uploading |-->| uploaded |
+ *       `------`     `---------`     `--------`
+ *          ^              |    .-------.
+ *          |              |`->| aborted |
+ *          |              |    `-------`
+ *          |  .------.    |    .---------.
+ *          `-| failed |<-` `->| timed_out |-.
+ *          |  `------`         `---------`  |
+ *          `-------------------------------`
+ * ```
+ */
+export enum FileState {
+  Queued = 'queued',
+  Uploading = 'uploading',
+  TimedOut = 'timed_out',
+  Aborted = 'aborted',
+  Uploaded = 'uploaded',
+  Failed = 'failed',
+}
+
+export enum FileSource {
+  /**
+   * the file is created using the native file picker
+   */
+  Browse = 'browse',
+
+  /**
+   * the file was created using drag and drop from their desktop
+   */
+  DragAndDrop = 'drag-and-drop',
+
+  /**
+   * the file was created by dragging the file from another webpage
+   */
+  Web = 'web',
+
+  /**
+   * the file is created from a data URL using the `fromDataURL`
+   * method for files. This usually means that the file was created
+   * manually by the developer on behalf of the user
+   */
+  DataUrl = 'data-url',
+
+  /**
+   * the file is created from a blob using the `fromBlob`
+   * method for files. This usually means that the file was created
+   * manually by the developer
+   */
+  Blob = 'blob',
+}
+
+/**
+ * Files provide a uniform interface for interacting
+ * with data that can be uploaded or read.
+ */
+export default class File {
+  // @TODO: Extract this "special" blob thingy into its own type
+  #blob: Blob & {
+    name?: string;
+  };
+  #source: FileSource;
+
+  queue?: Queue;
+
+  private constructor(blob: Blob, source: FileSource) {
+    this.#blob = blob;
+    this.#source = source;
+  }
+
+  /**
+   * The source of the file. This is useful
+   * for applications that want to gather
+   * analytics about how users upload their
+   * content.
+   */
+  get source(): FileSource {
+    return this.#source;
+  }
+
+  #id = `file-${uuid()}`;
+
+  /** A unique id generated for this file. */
+  get id(): string {
+    return this.#id;
+  }
+
+  /** The file name */
+  get name(): string {
+    // @TODO: this "default" name here looks wrong - the tests wanted it ?!?
+    return this.#blob?.name ?? 'blob';
```

Comment:
I added this default name `blob` here. It seemed necessary because tests demanded it, but I have no idea where else it was coming from?

I'd rather have no `blob` default name and change the tests? Dunno if it'll break some consumers app.

Reply:
Right, I see the problem here.

I've renamed the internal `File` class to `UploadFile` – been meaning to do this because it conflicts with native js/browser `File` class.

This has allowed me to refactor the Blob/File handling throughout the internals. Please [see my commit here](https://github.com/adopted-ember-addons/ember-file-upload/commit/cab88eb5f62394d1aac98a25671586f78e8af18e) which you could cherry-pick if you like.

Essentially there had been some shortcuts taken and overuse of the `fromBlob` factory. I've refactored so that the `UploadFile` instance always has a native `File` which makes the internals much simpler. Hope this helps.

Reply:
Aaaah, very nice addition. When I realized I'm in conflict with `File` then I aliased this one to `UploadFile`, too already. Very much appreciated change 👍 

And also answers my two questions, thank you :)